### PR TITLE
log full stack traces on error

### PR DESCRIPTION
**Pull request by dcrissman**

### CODE REVIEW

**Comment by dcrissman** [comment]
A stacktrace would be very useful for a bug I am tracking down.

@bserdar, @jewzaam, @paterczm  - This is ready for review.

**Comment by jewzaam** [comment]
@dcrissman what is the impact of throwing a runtime exception?  The concern I have is that it will get back to the client.  We do not want to have a request fail because a hook fails, this is a bad user experience.  It will be confusing.  Does the hook processing ignore errors thrown by hook implementations?

**Comment by dcrissman** [comment]
>  The concern I have is that it will get back to the client.

The Mediator wraps all CRUD operations in a try/catch and handles accordingly. That said I should probably use Error.get instead.

>  We do not want to have a request fail because a hook fails, this is a bad user experience.

I would say it depends on the hook. In the case of the esb hook, if an event was missed I would think that would be pretty bad.

**Comment by jewzaam** [comment]
Clients don't care about integration implications.  They just want their
data request to occur.  If the hook is firing it means their data has been
persisted.  Telling them a hook failed has no value to the client.  It
should be logged and ideally monitored.  I think of it in terms of a
relational database.  If you have a trigger on a table, the client knows
nothing about it.  If the trigger gets into an invalid / broken state
operations against that table fail.  From an operations point of view, the
quickest fix is to disable the trigger if it can't be quickly fixed.  We
have an opportunity with lightblue hooks to not impact the client when
something in a hook is broken.  Ergo, don't throw unchecked exceptions.

File: publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java [hunk 1]
```diff
@@ -133,25 +127,29 @@ public void processHook(EntityMetadata entityMetadata, HookConfiguration hookCon
...
+        try {
+            r = lightblueFactory.getMediator().insert(ireq);
+        } catch (ClassNotFoundException | IllegalAccessException | InvocationTargetException | NoSuchMethodException | InstantiationException | IOException e) {
+            throw Error.get(HOOK_NAME + ":MediatorException", e);
+        }
         if (!r.getErrors().isEmpty()) {
-            // TODO Better Handle Exception
             for (Error e : r.getErrors()) {
-                LOGGER.error(e.toString());
+                LOGGER.error("Lightblue Error", e);
```
Comment (bserdar) [inline publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java]: A more descriptive error message would be useful here. Same for the next one,

File: publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java [hunk 2]
```diff
@@ -133,25 +127,29 @@ public void processHook(EntityMetadata entityMetadata, HookConfiguration hookCon
...
         ArrayNode data = jsonNode.putArray("data");
         data.add(new ObjectMapper().valueToTree(entity));
 
         InsertionRequest ireq = InsertionRequest.fromJson(jsonNode);
-        Response r = lightblueFactory.getMediator().insert(ireq);
+        Response r;
+        try {
+            r = lightblueFactory.getMediator().insert(ireq);
+        } catch (ClassNotFoundException | IllegalAccessException | InvocationTargetException | NoSuchMethodException | InstantiationException | IOException e) {
+            throw Error.get(HOOK_NAME + ":MediatorException", e);
```
Comment (bserdar) [inline publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java]: This means there is a configuration error. You can throw a runtime exception for this one.

File: publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java (hunk 1 above)
Comment (bserdar) [inline publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java]: Maybe we can use a different logger, and send the error logs to a separate logfile?

File: publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java [hunk 3]
```diff
@@ -133,25 +127,29 @@ public void processHook(EntityMetadata entityMetadata, HookConfiguration hookCon
...
+        }
         if (!r.getErrors().isEmpty()) {
-            // TODO Better Handle Exception
             for (Error e : r.getErrors()) {
-                LOGGER.error(e.toString());
+                LOGGER.error("Lightblue Error", e);
             }
+            throw Error.get(HOOK_NAME + ":Error", "Errors while attempting to insert esb events.");
```
Comment (bserdar) [inline publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java]: Agree with Naveen. Log errors, don't throw.

Comment (jewzaam) [inline publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java]: @bserdar does the hook framework protect against implementations throwing
unchecked exception?

Comment (bserdar) [inline publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java]: Mediator catches all exceptions at the end, and anything that's not handled
ends up at the client side as a generic error.

**Comment by dcrissman** [comment]
@bserdar, @jewzaam - Requested changes made.

File: publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java (hunk 3 above)
Comment (jewzaam) [inline publish-hook/src/main/java/com/redhat/lightblue/hook/publish/PublishHook.java]: Created https://github.com/lightblue-platform/lightblue-core/issues/486

**Pull request merged**